contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,158 @@
+#include <eosio.system/eosio.system.hpp>
+
+#include <eosio/eosio.hpp>
+#include <eosio/privileged.hpp>
+#include <eosio/crypto_bls_ext.hpp>
+
+namespace eosiosystem {
+
+   bool system_contract::is_savanna_consensus() {
+      return !_gstate5.last_finalizers.empty();
+   }
+
+   bool system_contract::is_registered_producer( const name& producer ) {
+      // Returns true if `producer` is a registered producer
+   }
+
+   bool system_contract::is_active_producer( const name& producer ) {
+      // Returns true if `producer` is an active producer (top 21)
+   }
+
+   bool set_next_finalizers_set(const std::vector<name>& producers ) {
+      // Build finalizer set
+      std::vector<finalizer_authority> finalizers;
+      for ( const auto p&: producers ) {
+         auto it = _finalizer_keys.find( p.value );
+         if( it != _finalizer_keys.end() ) { // p has registered a finalizer key
+            finalizers.emplace_back(
+               finalizer_authority{
+                  .description = "",
+                  .weight      = 1,
+                  .public_key  = it->active_key
+               }
+            );
+         }
+      }
+
+      // Abort if less than 15 top producers have registered a finalizer key
```

Comment:
We need to be willing to go down beyond the top 21 potentially to find exactly 21 BPs that meet the requirements.

Perhaps we can limit it to the top 30. Out of the top 30, we should find 21 BPs that meet all the normal requirements for being a proposer AND also have an non-duplicate active finalizer key.